### PR TITLE
SONARPY-525 Rule S5655: Arguments given to functions should be of an expected type

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 1]
```diff
@@ -180,6 +180,10 @@ private void createAmbiguousSymbols() {
         case FUNC_DECLARATION:
           FunctionDef functionDef = (FunctionDef) bindingUsage.tree().parent();
           FunctionSymbolImpl functionSymbol = new FunctionSymbolImpl(functionDef, symbol.fullyQualifiedName(), pythonFile);
+          ParameterList parameters = functionDef.parameters();
```
Comment (andrea-guarino-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java:183]: Why can't we do it inside `FunctionSymbolImpl` constructor anymore?

File: python-frontend/src/main/java/org/sonar/python/semantic/FunctionSymbolImpl.java [hunk 2]
```diff
@@ -114,7 +120,34 @@ private static boolean isInstanceMethod(FunctionDef functionDef) {
         List<Name> names = decorator.name().names();
         return names.get(names.size() - 1).name();
       })
-      .noneMatch(decorator -> decorator.equals("staticmethod") || decorator.equals("classmethod"));
+      .noneMatch(decorator -> decorator.equals(STATIC_METHOD_DECORATOR) || decorator.equals(CLASS_METHOD_DECORATOR));
+  }
+
+  private static HasFirstImplicitParam hasImplicitFirstParameter(FunctionDef functionDef) {
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/FunctionSymbolImpl.java]: I have removed that API and added a new one to retrieve decorators in d9e48824b4d75f149e2db5cfcac09b46ad1a8fe9.

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 3]
```diff
@@ -170,9 +173,27 @@ public static ClassSymbol typeShedClass(String fullyQualifiedName) {
...
     }
+
+    private static void setParameterTypes(Symbol symbol, FunctionDef functionDef) {
+      if (symbol.is(Symbol.Kind.FUNCTION)) {
+        FunctionSymbolImpl functionSymbol = (FunctionSymbolImpl) symbol;
+        ParameterList parameters = functionDef.parameters();
+        if (parameters != null) {
+          // For builtin functions, we don't have type information from typings.pyi for the parameters when constructing the initial symbol table
+          // We need to recreate those with that information
+          functionSymbol.parameters().clear();
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java]: Indeed! I refactored that part in d9e48824b4d75f149e2db5cfcac09b46ad1a8fe9.

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java (hunk 1 above)
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: I tried adding that logic again in `FunctionSymbolImpl` constructor but encountered a different issue where some class symbols could have their super class read through type annotations before their type hierarchy was defined.
Therefore, in d9e48824b4d75f149e2db5cfcac09b46ad1a8fe9, I moved this logic to `ThirdPhaseVisitor`.

File: python-checks/src/test/resources/checks/argumentType.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
+from math import acos
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-checks/src/test/resources/checks/argumentType.py:1]: Indeed! Actually that made me realize we don't handle them very well as we only rely on type annotation without knowing whether the parameter is variadic or not. In 6a53bddec8101c389dc573bcbf56eb2f79c5a585 I have therefore disabled the rule when a function has variadic parameters presents.

File: python-frontend/src/main/java/org/sonar/python/semantic/FunctionSymbolImpl.java [hunk 5]
```diff
@@ -201,13 +230,15 @@ public void setOwner(Symbol owner) {
...
 
     private final String name;
+    private final InferredType declaredType;
     private final boolean hasDefaultValue;
     private final boolean isKeywordOnly;
     private final boolean isPositionalOnly;
     private final LocationInFile location;
 
-    ParameterImpl(@Nullable String name, boolean hasDefaultValue, ParameterState parameterState, @Nullable LocationInFile location) {
+    ParameterImpl(@Nullable String name, @Nullable InferredType declaredType, boolean hasDefaultValue, ParameterState parameterState, @Nullable LocationInFile location) {
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/FunctionSymbolImpl.java:239]: I forgot to remove that...Fixed it in 6a53bddec8101c389dc573bcbf56eb2f79c5a585

File: python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java [hunk 6]
```diff
@@ -71,6 +71,19 @@ public boolean canBeOrExtend(String typeName) {
     return typeClass.canBeOrExtend(typeName);
   }
 
+  @Override
+  public boolean canBeOrExtend(InferredType other) {
+    if (other instanceof RuntimeType) {
+      String otherFullyQualifiedName = ((RuntimeType) other).typeClass.fullyQualifiedName();
+      return otherFullyQualifiedName != null ? this.canBeOrExtend(otherFullyQualifiedName) : typeClass.hasUnresolvedTypeHierarchy();
```
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java]: Indeed, also this case should normally never happen.
I fixed this in 6a53bddec8101c389dc573bcbf56eb2f79c5a585.
Please note that b275a488d001987df20f088dc8a1711c6ffe47fe also introduce changes to this method to account for duck type compatibility.

**Pull request merged**